src/main/scala/com.snowplowanalytics.iglu/client/resolver/registries/RegistryLookup.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2014-2018 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.iglu.client.resolver.registries
+
+// Java
+import java.net.UnknownHostException
+
+// Scala
+import scala.util.control.NonFatal
+
+// cats
+import cats.Id
+import cats.data.{EitherT, OptionT}
+import cats.effect.{IO, Sync}
+import cats.implicits._
+import cats.effect.implicits._
+
+// circe
+import io.circe.Json
+import io.circe.parser.parse
+
+// Iglu Core
+import com.snowplowanalytics.iglu.core.{SchemaKey, SelfDescribingSchema}
+import com.snowplowanalytics.iglu.core.circe.implicits._
+
+/** A capability of `F` to perform a schema lookup action, using `RepositoryRef` ADT
+ *
+ * @tparam F effect type
+ */
+trait RegistryLookup[F[_]] {
+  // This can be abstracted over RepositoryRef to let others extend functionality of Resolver
+
+  /** Find a schema in the particular `RepositoryRef`
+   *
+   * @param repositoryRef one of supported repository types
+   * @param schemaKey The SchemaKey uniquely identifying the schema in Iglu
+   * @return either schema parsed into `Json` or `RegistryError`, such as absent schema,
+   *         or unexpected response
+   */
+  def lookup(repositoryRef: Registry, schemaKey: SchemaKey): F[Either[RegistryError, Json]]
+}
+
+object RegistryLookup {
+
+  def apply[F[_]](implicit ev: RegistryLookup[F]): RegistryLookup[F] = ev
+
+  implicit class LookupOps(repositoryRef: Registry) {
+    def lookupSchema[F[_]: RegistryLookup](schemaKey: SchemaKey): F[Either[RegistryError, Json]] =
+      RegistryLookup[F].lookup(repositoryRef, schemaKey)
+  }
+
+  implicit def ioLookupInstance[F[_]](implicit F: Sync[F]): RegistryLookup[F] =
+    new RegistryLookup[F] {
+      def lookup(repositoryRef: Registry, schemaKey: SchemaKey): F[Either[RegistryError, Json]] =
+        repositoryRef match {
+          case Registry.Http(_, connection)  => httpLookup(connection, schemaKey)
+          case Registry.Embedded(_, path)    => embeddedLookup[F](path, schemaKey)
+          case Registry.InMemory(_, schemas) => F.pure(inMemoryLookup(schemas, schemaKey))
+        }
+    }
+
+  // Id instance also swallows all exceptions into `RegistryError`
+  implicit def idLookupInstance: RegistryLookup[Id] = new RegistryLookup[Id] {
```

Comment:
could we have a new one with `Eval`?

Reply:
Yep, will switch everything to `Eval` today

Reply:
@BenFradet, `M5` with `RegistryLookup[Eval]` will be on its way to Maven after `scala-lru-map:0.3.0-M3` is published.